### PR TITLE
Splitstore: support on-disk marksets using badger

### PR DESCRIPTION
**Pull request by vyzo**
Very useful if you run out of memory during compaction.

### CODE REVIEW

**Comment by vyzo** [comment]
Notes:
- marking is about twice as slow than with the in-memory map, but we have no memory pressure.
- block validation time is unaffected, thanks to the `FileIO` badger options in the markset.
- removed the markset occurs check from `trackTxnRefs` so as not to hit the markset; that's a non-critical optimization that doesn't really affect anything.
- added some documentation for configuration options in the README.

File: blockstore/splitstore/README.md [hunk 1]
```diff
@@ -27,9 +27,34 @@ If you intend to use the discard coldstore, your also need to add the following:
     ColdStoreType = "discard"
```
Comment (raulk) [inline blockstore/splitstore/README.md:27]: Can we add a big warning saying that the usage of `discard` is discouraged unless you absolutely know what you're doing?

File: blockstore/splitstore/README.md [hunk 2]
```diff
@@ -27,9 +27,34 @@ If you intend to use the discard coldstore, your also need to add the following:
...
+
+- `HotStoreType` -- specifies the type of hotstore to use.
+  The only currently supported option is `"badger"`.
+- `ColdStoreType` -- specifies the type of coldstore to use.
+  The default value is `"universal"`, which will use the initial monolith blockstore
+  as the coldstore.
+  The other possible value is `"discard"`, as outlined above, which is specialized for
+  running without a coldstore. Note that the discard store wraps the initial monolith
+  blockstore and discards writes; this is necessary to support syncing from a snapshot.
+- `MarkSetType` -- specifies the type of markset to use during compaction.
```
Comment (raulk) [inline blockstore/splitstore/README.md:47]: Have we defined what a "markset" is? If not, it makes this hard to follow.

File: blockstore/splitstore/markset_badger.go [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+	if err != nil {
+		return nil, xerrors.Errorf("error creating markset directory: %w", err)
+	}
+
+	opts := badger.DefaultOptions(path)
+	opts.SyncWrites = false
+	opts.CompactL0OnClose = false
+	opts.Compression = options.None
+	opts.TableLoadingMode = options.FileIO
+	opts.ValueLogLoadingMode = options.FileIO
```
Comment (raulk) [inline blockstore/splitstore/markset_badger.go:68]: Can we add a comment for our future selves here, on why we're using `FileIO`?

File: blockstore/splitstore/markset_badger.go [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+	err := os.MkdirAll(msPath, 0755) //nolint:gosec
+	if err != nil {
+		return nil, xerrors.Errorf("error creating markset directory: %w", err)
+	}
+
+	return &BadgerMarkSetEnv{path: msPath}, nil
+}
+
+func (e *BadgerMarkSetEnv) Create(name string, sizeHint int64) (MarkSet, error) {
+	path := filepath.Join(e.path, name)
```
Comment (raulk) [inline blockstore/splitstore/markset_badger.go:50]: Do we need to add a salt to this filename on the caller side? `Create` always get called with a fixed name `live`. This means that:

1. on restart, we will open an _existing_ markset. However, this is unsafe because `SyncWrites` below is false!
2. a failure to `Close` and `Remove` will end up opening a previously existing database.

If we always want to operate on the same database, we need to ensure that reopening it will always be safe. However, there's probably little value in doing that because we do not resume marking on a process restart. So the recommendation here is to add a salt.

OTOH, an interruption could lead to leftover/abandoned markset databases.

File: blockstore/splitstore/markset_badger.go [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+		return nil, xerrors.Errorf("error clearing markset directory: %w", err)
+	}
+
+	err = os.MkdirAll(path, 0755) //nolint:gosec
+	if err != nil {
+		return nil, xerrors.Errorf("error creating markset directory: %w", err)
+	}
+
+	opts := badger.DefaultOptions(path)
+	opts.SyncWrites = false
```
Comment (raulk) [inline blockstore/splitstore/markset_badger.go:64]: Consider flipping to `true` depending on the above.

File: blockstore/splitstore/markset_badger.go [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+
+	err := db.Close()
+	if err != nil {
+		return xerrors.Errorf("error closing badger markset: %w", err)
+	}
+
+	err = os.RemoveAll(s.path)
+	if err != nil {
+		return xerrors.Errorf("error deleting badger markset: %w", err)
+	}
```
Comment (raulk) [inline blockstore/splitstore/markset_badger.go:210]: Close is not idempotent. If called multiple times or concurrently, only the caller that gets to execute the close logic will see any errors. Subsequent callers would not see this error. Consider storing a `closeErr` and returning that when `s.pend == nil` above.

File: blockstore/splitstore/splitstore_compact.go [hunk 7]
```diff
@@ -184,16 +184,6 @@ func (s *SplitStore) trackTxnRef(c cid.Cid) {
...
-	if s.txnProtect != nil {
-		mark, err := s.txnProtect.Has(c)
-		if err != nil {
-			log.Warnf("error checking markset: %s", err)
-			// track it anyways
-		} else if mark {
-			return
-		}
-	}
-
```
Comment (raulk) [inline blockstore/splitstore/splitstore_compact.go:196]: Why remove this?

File: blockstore/splitstore/splitstore_compact.go [hunk 8]
```diff
@@ -209,27 +199,11 @@ func (s *SplitStore) trackTxnRefMany(cids []cid.Cid) {
...
-					log.Warnf("error checking markset: %s", err)
-				}
-				// track it anyways
-			}
-
-			if mark {
-				continue
-			}
-		}
-
```
Comment (raulk) [inline blockstore/splitstore/splitstore_compact.go:232]: Why remove this?

File: blockstore/splitstore/README.md (hunk 1 above)
Comment (vyzo) [reply blockstore/splitstore/README.md:27]: well, it's not exactly discouraged, but you should definitely know what you are doing!
I'll add something along this lines.

File: blockstore/splitstore/README.md (hunk 2 above)
Comment (vyzo) [reply blockstore/splitstore/README.md:47]: Sure, I'll define it.

File: blockstore/splitstore/markset_badger.go (hunk 3 above)
Comment (vyzo) [reply blockstore/splitstore/markset_badger.go:68]: Yeah, good call -- this is kind of critical.

File: blockstore/splitstore/markset_badger.go (hunk 4 above)
Comment (vyzo) [reply blockstore/splitstore/markset_badger.go:50]: No, on restart we just delete everything -- marksets are transient.

File: blockstore/splitstore/markset_badger.go (hunk 5 above)
Comment (vyzo) [reply blockstore/splitstore/markset_badger.go:64]: Nope, not needed -- the existing markset is deleted on restart.

File: blockstore/splitstore/markset_badger.go (hunk 6 above)
Comment (vyzo) [reply blockstore/splitstore/markset_badger.go:210]: Maybe, but It doesn't really matter, it's a throwaway datastore!

File: blockstore/splitstore/splitstore_compact.go (hunk 7 above)
Comment (vyzo) [reply blockstore/splitstore/splitstore_compact.go:196]: so that we don't hit the markset and potentially stall an I/O operation.

File: blockstore/splitstore/splitstore_compact.go (hunk 8 above)
Comment (vyzo) [reply blockstore/splitstore/splitstore_compact.go:232]: same as above.

File: blockstore/splitstore/markset_badger.go (hunk 5 above)
Comment (vyzo) [reply blockstore/splitstore/markset_badger.go:64]: Actually when we reopen it.

File: blockstore/splitstore/markset_badger.go (hunk 4 above)
Comment (vyzo) [reply blockstore/splitstore/markset_badger.go:50]: Correction: when we reopen the markset, it's deleted.

So basically there is no safety concerns from close errors. The close implementation deletes the thing, and when we reopen it we delete again just to be sure we didn't have left overs.

Basically the on-disk markset is completely transient, we don't care at all about persisting it once we are done using it.

File: blockstore/splitstore/README.md (hunk 1 above)
Comment (vyzo) [reply blockstore/splitstore/README.md:27]: Added a warning.

File: blockstore/splitstore/README.md (hunk 2 above)
Comment (vyzo) [reply blockstore/splitstore/README.md:47]: Added a definition.

File: blockstore/splitstore/markset_badger.go (hunk 4 above)
Comment (vyzo) [reply blockstore/splitstore/markset_badger.go:50]: On the other hand, using a salt would leave us with leftover marksets we can't consistently delete on restart, basically garbage on disk.
So using a salt is not the right thing to do here, and the code behaves as desired afaict.

File: blockstore/splitstore/markset_badger.go (hunk 5 above)
Comment (vyzo) [reply blockstore/splitstore/markset_badger.go:64]: Also, it would slow down things significantly.

File: blockstore/splitstore/markset_badger.go (hunk 3 above)
Comment (vyzo) [reply blockstore/splitstore/markset_badger.go:68]: done.

File: blockstore/splitstore/markset_badger.go (hunk 4 above)
Comment (raulk) [reply blockstore/splitstore/markset_badger.go:50]: Ack. Makes sense.

File: blockstore/splitstore/markset_badger.go (hunk 5 above)
Comment (raulk) [reply blockstore/splitstore/markset_badger.go:64]: Ack.

File: blockstore/splitstore/markset_badger.go (hunk 6 above)
Comment (raulk) [reply blockstore/splitstore/markset_badger.go:215]: Happy to go this way.

**Review by raulk — APPROVED** [review]

**Review by magik6k — APPROVED** [review]
